cloudify/decorators.py

```diff
@@ -40,4 +45,87 @@ def workflow(func=None, system_wide=False, resumable=False, **kwargs):
+    states = states if isinstance(
+        states, list) else ['started']
+    rest_client = get_rest_client()
+
+    def outer_wrapper(func):
+        def inner_wrapper(*args, **kwargs):
+            _ctx = kwargs.get('ctx')
+            if _ctx.workflow_id in workflows:
+                # Split the instances at the index into
+                # "priority_instances" and "no_priority"
+                # node instances. "priority_instances means that these
+                # instances' index is a smaller number than the
+                # current node instances's. They should be executed first.
+                # "no_priority" means that it comes after.
+                priority_instances, no_priority = \
+                    split_instances_at_index(rest_client)
+                # Finished instances are those that are already in the
+                # state where they are not considered blockers of our
+                # current node instance.
+                non_blocking_instances = [
+                    i for i in priority_instances + no_priority
+                    if i.state in states]
+                try:
+                    # preceding_instance is the instance directly preceding
+                    # the current node instance according to index.
+                    preceding_instance = priority_instances[-1]
+                except IndexError:
+                    # If there are no priority_instances instances, we can just
+                    # run the operation.
+                    pass
+                else:
+                    # We want to wait if the following conditions exist:
+                    # 1) revert_to_parallel_threshold is 0 and preceding
+                    #    node instance (preceding_instance) is not ready.
+                    #    This means - always try to execute pure serial.
+                    # 2) revert_to_parallel_threshold is unequal to total
+                    #    finished instances. This means that if we have
+                    #    exceeded the revert_to_parallel_threshold threshold,
+                    #    then we would return to parallel execution.
+                    if (revert_to_parallel_threshold == 0 or
+                        revert_to_parallel_threshold != len(
+                                non_blocking_instances)) and \
+                            preceding_instance and \
+                            preceding_instance.state not in states:
+                        raise CloudifySerializationRetry(
+                            'Serial Cloudify Operation. '
+                            'Waiting for the following '
+                            'node instances to be ready: '
+                            '{priority_instances}.'.format(
+                                priority_instances=[(p.id, p.state) for p in
+                                                    priority_instances]))
+            return operation(func, *args, **kwargs)
```

Comment:
this needs to eventually call the function
a decorator can be written like this

```python
def serial_operation(...):
  @functools.wraps
  def outer(f):
    def inner(*a, **kw):
      return f(*a, **kw)
    return operation(func=inner)
  return outer
```

Reply:
also remember to pass the decorator arguments to `operation`'s kwargs (so that you can `@serial_operation(resumable=True)`

Reply:
You mean I need to add that in the serial_operation args?